### ISSUE TITLE
Casting error  

### ISSUE THREAD

**Issue by zufus** [numba/numba#9461]
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
  visible in the release notes
  (https://numba.readthedocs.io/en/stable/release-notes-overview.html).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

Hi, I am attaching a bug.py script to show a bug I encountered: basically it seems that by calling "np.astype()" to cast an array from float to unsigned integers within a function compiled with njit creates an error.

I am using python 3.10.5 64-bit on Windows 10.

The attached script calls the same function twice: once in the numba accellerated version, and once with a classical version, showing that there are two different results.

The bug is solved by moving the cast operation outside the njit compiled function.

Thanks,

bug.py:
```
import numpy as np 
from matplotlib import pyplot as plt
from numba import njit

@njit(parallel = True)
def makeRampsFast(period, x):
    return (256 / period *x).astype(np.ubyte)
    
def makeRampsSlow(period, x):
    return (256 / period *x).astype(np.ubyte)
    
period = 100
x, y = np.mgrid[0:1024, 0:1024]


ramps_f = makeRampsFast(period, x)
ramps_s = makeRampsSlow(period, x)

if np.abs((ramps_f - ramps_s)).sum() != 0:
    print("Houston, we've had a bug here!")

fig, (ax1, ax2) = plt.subplots(1,2)


ax1.matshow(ramps_s, aspect='equal', cmap='Greys')
ax2.matshow(ramps_f, aspect='equal', cmap='Greys')

ax1.set_title('Correct output - w/o numba')
ax2.set_title('Wrong output - w/ numba')

#ax.view_init(90, 90, 0)


plt.show()
```

**Comment by sklam** [comment]
Thanks for the bug report. 

There's a odd problem as the result depends on the size of the array.

```python
import numpy as np
from numba import njit

@njit
def trunc(x):
    return x.astype(np.ubyte)


print("TEST 1".center(80, '-'))
arr = np.asarray([2618.88] * 2)
print(arr)
print(arr.astype(np.ubyte))
print(trunc(arr))

print("TEST 2".center(80, '-'))
arr = np.asarray([2618.88] * 4)
print(arr)
print(arr.astype(np.ubyte))
print(trunc(arr))
```
prints:

```
-------------------------------------TEST 1-------------------------------------
[2618.88 2618.88]
[58 58]
[58 58]
-------------------------------------TEST 2-------------------------------------
[2618.88 2618.88 2618.88 2618.88]
[58 58 58 58]
[255 255 255 255]
```

**Comment by sklam** [comment]
TL;DR

This is triggering a undefined behavior in both NumPy and Numba. Use `np.clip` to ensure values are within [0, 255] to avoid problems.

---

After consulting the C standard and the LLVM spec (which is more or less following C semantic), the reproducer will trigger a undefined behavior in both NumPy's C code and Numba's LLVM JIT code. The `.astype(np.ubyte)` is implemented as a simple assignment to slot pointed by `unsigned byte*` in both NumPy and Numba. When the values being type-casted is outside of the range of the destination type, it can result in undefined behavior. 

```
In [2]: import numpy as np

In [3]: x, y = np.mgrid[0:1024, 0:1024]

In [4]: x
Out[4]:
array([[   0,    0,    0, ...,    0,    0,    0],
       [   1,    1,    1, ...,    1,    1,    1],
       [   2,    2,    2, ...,    2,    2,    2],
       ...,
       [1021, 1021, 1021, ..., 1021, 1021, 1021],
       [1022, 1022, 1022, ..., 1022, 1022, 1022],
       [1023, 1023, 1023, ..., 1023, 1023, 1023]])

In [5]: period = 100

In [6]: (256 / period * x)
Out[6]:
array([[0.00000e+00, 0.00000e+00, 0.00000e+00, ..., 0.00000e+00,
        0.00000e+00, 0.00000e+00],
       [2.56000e+00, 2.56000e+00, 2.56000e+00, ..., 2.56000e+00,
        2.56000e+00, 2.56000e+00],
       [5.12000e+00, 5.12000e+00, 5.12000e+00, ..., 5.12000e+00,
        5.12000e+00, 5.12000e+00],
       ...,
       [2.61376e+03, 2.61376e+03, 2.61376e+03, ..., 2.61376e+03,
        2.61376e+03, 2.61376e+03],
       [2.61632e+03, 2.61632e+03, 2.61632e+03, ..., 2.61632e+03,
        2.61632e+03, 2.61632e+03],
       [2.61888e+03, 2.61888e+03, 2.61888e+03, ..., 2.61888e+03,
        2.61888e+03, 2.61888e+03]])
```

`Out[6]` shows that the array before `.astype(np.ubyte)` has values outside of [0, 255]. To fix this, we need to clip the values:

```python
@njit(parallel=True)
def makeRampsFast(period, x):
    return np.clip((256 / period *x), 0, 255).astype(np.ubyte)

def makeRampsSlow(period, x):
    return np.clip((256 / period *x), 0, 255).astype(np.ubyte)

```

Ensuring the values fall within [0, 255] will avoid undefined behavior. Numba/LLVM was switching between directly taking the lower 8-bytes of the truncated double, or saturating it due to use of `packuswb` x86 instruction in the SIMD path. Even NumPy's behavior is not guaranteed to be the same across platforms and the compilers used to build the package.

Here, Numba is following the same semantic (or the lack of) as NumPy for casting out-of-range values. There is no action for Numba.


For C-std and LLVM spec details, see 
- https://wiki.sei.cmu.edu/confluence/display/c/FLP34-C.+Ensure+that+floating-point+conversions+are+within+range+of+the+new+type 
- https://c0x.shape-of-code.com/6.3.1.4.html 
- https://llvm.org/docs/LangRef.html#id266

For details of NumPy astype implementation, see: https://github.com/numpy/numpy/blob/3b373f1f58a68e898cd67ccbeeab9ea17b7978a7/numpy/_core/src/multiarray/methods.c#L865

**Comment by zufus** [comment]
@sklam thanks for the reply, but I don't think it fixes the issue. I am using np.astype(ubyte) exactly to wrap the function to the (0,255) interval, transforming a linear function into a periodic ramp function. 
If I clip the values to (0,255) BEFORE calling np.astype(ubyte), I obtain the wrong result.
The function makeRampsSlow, called outside Numba works flawlessly everywhere. 

So I think it is an issue in Numba.

Thanks

**Comment by zufus** [comment]
Hi, thanks! I think the bug may be closed!

**Comment by stuartarchibald** [comment]
Thanks for confirming @zufus, closing.
